kaldb/src/main/java/com/slack/kaldb/blobfs/s3/S3CrtBlobFs.java

```diff
@@ -0,0 +1,548 @@
+    } catch (InterruptedException | ExecutionException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private boolean isPathTerminatedByDelimiter(URI uri) {
+    return uri.getPath().endsWith(DELIMITER);
+  }
+
+  private String normalizeToDirectoryPrefix(URI uri) throws IOException {
+    Preconditions.checkNotNull(uri, "uri is null");
+    URI strippedUri = getBase(uri).relativize(uri);
+    if (isPathTerminatedByDelimiter(strippedUri)) {
+      return sanitizePath(strippedUri.getPath());
+    }
+    return sanitizePath(strippedUri.getPath() + DELIMITER);
+  }
+
+  private URI normalizeToDirectoryUri(URI uri) throws IOException {
+    if (isPathTerminatedByDelimiter(uri)) {
+      return uri;
+    }
+    try {
+      return new URI(uri.getScheme(), uri.getHost(), sanitizePath(uri.getPath() + DELIMITER), null);
+    } catch (URISyntaxException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private String sanitizePath(String path) {
+    path = path.replaceAll(DELIMITER + "+", DELIMITER);
+    if (path.startsWith(DELIMITER) && !path.equals(DELIMITER)) {
+      path = path.substring(1);
+    }
+    return path;
+  }
+
+  private URI getBase(URI uri) throws IOException {
+    try {
+      return new URI(uri.getScheme(), uri.getHost(), null, null);
+    } catch (URISyntaxException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private boolean existsFile(URI uri) throws IOException {
+    try {
+      URI base = getBase(uri);
+      String path = sanitizePath(base.relativize(uri).getPath());
+      HeadObjectRequest headObjectRequest =
+          HeadObjectRequest.builder().bucket(uri.getHost()).key(path).build();
+
+      s3AsyncClient.headObject(headObjectRequest).get();
+      return true;
+    } catch (Exception e) {
+      if (e instanceof ExecutionException && e.getCause() instanceof NoSuchKeyException) {
+        return false;
+      } else {
+        throw new IOException(e);
+      }
+    }
+  }
+
+  private boolean isEmptyDirectory(URI uri) throws IOException {
+    if (!isDirectory(uri)) {
+      return false;
+    }
+    String prefix = normalizeToDirectoryPrefix(uri);
+    boolean isEmpty = true;
+    ListObjectsV2Response listObjectsV2Response;
+    ListObjectsV2Request.Builder listObjectsV2RequestBuilder =
+        ListObjectsV2Request.builder().bucket(uri.getHost());
+
+    if (!prefix.equals(DELIMITER)) {
+      listObjectsV2RequestBuilder = listObjectsV2RequestBuilder.prefix(prefix);
+    }
+
+    ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.build();
+    try {
+      listObjectsV2Response = s3AsyncClient.listObjectsV2(listObjectsV2Request).get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new IOException(e);
+    }
+
+    for (S3Object s3Object : listObjectsV2Response.contents()) {
+      if (s3Object.key().equals(prefix)) {
+        continue;
+      } else {
+        isEmpty = false;
+        break;
+      }
+    }
+    return isEmpty;
```

Comment:
That would indeed make this a bit easier to follow instead of using the current `continue` behavior, but it does match the existing client here - https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/blobfs/s3/S3BlobFs.java#L198